libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/optimiser/NameDispenser.h>
+
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace {
+void appendMemoryStores(
+	vector<Statement>& _statements,
+	langutil::SourceLocation _loc,
+	vector<pair<YulString, YulString>> const& _mstores
+)
+{
+	for (auto const& [mpos, value]: _mstores)
+		_statements.emplace_back(ExpressionStatement{_loc, FunctionCall{
+			_loc,
+			Identifier{_loc, "mstore"_yulstring},
+			{
+				Literal{_loc, LiteralKind::Number, mpos, {}},
+				Identifier{_loc, value}
+			}
+		}});
+}
+}
+
+StackToMemoryMover::StackToMemoryMover(
+	OptimiserStepContext& _context,
+	map<YulString, map<YulString, u256>> const& _memoryOffsets
+): m_memoryOffsets(_memoryOffsets), m_nameDispenser(_context.dispenser)
+{
+}
+
+void StackToMemoryMover::operator()(FunctionDefinition& _functionDefinition)
+{
+	auto saved = m_currentFunctionMemoryOffsets;
```

Comment:
move? swap?

Reply:
Ok, you got me there: maybe I shouldn't use ``auto`` that much after all :-). It's a const pointer :-).